### PR TITLE
Adds user_id and presence logging to transport.address_mgr.

### LINKED ISSUE #4390: Address Mgr status info

### PR DESCRIPTION
**Pull request by err508**
Closes #4390.

### CODE REVIEW

File: raiden/network/transport/matrix/transport.py [hunk 1]
```diff
@@ -395,6 +395,7 @@ def on_success(greenlet):
         self._client.sync_thread.link_exception(self.on_error)
         self._client.sync_thread.link_value(on_success)
         self.greenlets = [self._client.sync_thread]
+        self._schedule(30, self._address_mgr.get_status_info())
```
Comment (ulope) [inline raiden/network/transport/matrix/transport.py]: You're calling the function instead of passing it.

File: raiden/network/transport/matrix/utils.py [hunk 2]
```diff
@@ -251,6 +251,22 @@ def _presence_listener(self, event: Dict[str, Any]):
...
+            }
+            for address in self.known_addresses
+        }
+
+        log.debug(
+            "Matrix status info",
+            current_user_id=self._user_id,
+            user_ids_and_presence_for_known_addresses=user_ids_to_presence_for_known_addresses,
+        )
+        return
```
Comment (ulope) [inline raiden/network/transport/matrix/utils.py]: Nitpick: the `return` is superfluous

File: raiden/network/transport/matrix/utils.py [hunk 3]
```diff
@@ -251,6 +251,22 @@ def _presence_listener(self, event: Dict[str, Any]):
...
+                user_id: self._userid_to_presence[user_id].value
+                for user_id in self._address_to_userids[address]
+            }
+            for address in self.known_addresses
+        }
+
+        log.debug(
+            "Matrix status info",
+            current_user_id=self._user_id,
+            user_ids_and_presence_for_known_addresses=user_ids_to_presence_for_known_addresses,
```
Comment (ulope) [inline raiden/network/transport/matrix/utils.py]: The variable is maybe a bit too detailed ;)

File: raiden/network/transport/matrix/utils.py [hunk 4]
```diff
@@ -251,6 +251,22 @@ def _presence_listener(self, event: Dict[str, Any]):
...
+        user_ids_to_presence_for_known_addresses = {
+            to_checksum_address(address): {
+                user_id: self._userid_to_presence[user_id].value
+                for user_id in self._address_to_userids[address]
+            }
+            for address in self.known_addresses
+        }
+
+        log.debug(
+            "Matrix status info",
```
Comment (ulope) [inline raiden/network/transport/matrix/utils.py]: It's not really general status, maybe call it `Matrix presence info`?

File: raiden/network/transport/matrix/transport.py [hunk 5]
```diff
@@ -1198,11 +1203,7 @@ def _get_room_ids_for_address(
         address_hex: AddressHex = to_checksum_address(address)
```
Comment (Dominik1999) [inline raiden/network/transport/matrix/transport.py]: is hex here the same as checksum? or should you rename `address_hex` to `address_checksum ?`

File: raiden/network/transport/matrix/transport.py [hunk 6]
```diff
@@ -470,6 +471,10 @@ def on_success(greenlet):
         self.greenlets.append(greenlet)
         return greenlet
 
+    def _schedule(self, period: float, func: Callable, *args, **kwargs):
+        gevent.spawn_later(0, func, *args, **kwargs)
+        gevent.spawn_later(period, self._schedule, period=period, func=func, *args, **kwargs)
```
Comment (Dominik1999) [inline raiden/network/transport/matrix/transport.py]: Can you explain what this function does? So what I understand is that it spawns `self._schedule` after the given `period`. But this spawn_later is called in the very same function? Just want to make sure, that this is not a loop

File: raiden/network/transport/matrix/transport.py [hunk 7]
```diff
@@ -1198,11 +1203,7 @@ def _get_room_ids_for_address(
         address_hex: AddressHex = to_checksum_address(address)
         with self._account_data_lock:
             room_ids = self._client.account_data.get("network.raiden.rooms", {}).get(address_hex)
-            self.log.debug(
-                "matrix get account data",
-                room_ids=room_ids,
-                for_address=to_checksum_address(address),
-            )
+            self.log.debug("Room ids for address", for_addres=address_hex, room_ids=room_ids)
```
Comment (Dominik1999) [inline raiden/network/transport/matrix/transport.py]: `for_addres` misses an additional **s**, right? maybe use address_checksum again

File: raiden/network/transport/matrix/utils.py [hunk 8]
```diff
@@ -251,6 +251,22 @@ def _presence_listener(self, event: Dict[str, Any]):
         if self._user_presence_changed_callback:
             self._user_presence_changed_callback(user, new_state)
 
+    def get_status_info(self):
```
Comment (Dominik1999) [inline raiden/network/transport/matrix/utils.py]: maybe call this function `get_matrix_status_info` ? Or which status are you interested in?

File: raiden/tests/integration/network/transport/test_matrix_transport.py [hunk 9]
```diff
@@ -128,7 +128,7 @@ def ping_pong_message_success(transport0, transport1):
     transport1._raiden_service.sign(pong_message)
     transport0.send_async(queueid1, ping_message)
 
-    with Timeout(20, exception=False):
+    with Timeout(40, exception=False):
```
Comment (Dominik1999) [inline raiden/tests/integration/network/transport/test_matrix_transport.py]: why did you change from 20 to 40?

File: raiden/network/transport/matrix/transport.py (hunk 1 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Right... didn't make a difference in this case as it has no args anyway right?

File: raiden/network/transport/matrix/transport.py (hunk 6 above)
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: Good point.  It is effectively an endless loop. 

This needs to stop when the transport stop event is set (or rather the AddressManager stop event, see #4333)

File: raiden/network/transport/matrix/transport.py (hunk 1 above)
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: No, it does! You're calling the function once here and then passing `None` to `_schedule()`

File: raiden/tests/integration/network/transport/test_matrix_transport.py (hunk 9 above)
Comment (err508) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: I think it's a tradeoff. For the CI long timeouts mean it might take long until tests fail, locally tests might fail because the timeouts are not long enough. I had this today so I changed it to not have test failure because my machine is to slow.

File: raiden/network/transport/matrix/transport.py (hunk 6 above)
Comment (Dominik1999) [reply raiden/network/transport/matrix/transport.py]: so it is intended to be an endless loop?

File: raiden/network/transport/matrix/transport.py (hunk 7 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: right!

File: raiden/network/transport/matrix/transport.py (hunk 6 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: I'll add a docstring, good point

File: raiden/network/transport/matrix/transport.py (hunk 5 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: to_checksum_address returns hex, so in a way yeah it's the same

File: raiden/network/transport/matrix/utils.py (hunk 4 above)
Comment (err508) [reply raiden/network/transport/matrix/utils.py]: `Address mager info`? It also logs known addresses, user_ids.

File: raiden/network/transport/matrix/utils.py (hunk 3 above)
Comment (err508) [reply raiden/network/transport/matrix/utils.py]: ooh not pythonic?:D Changed it `addresses_uids_presence`

File: raiden/network/transport/matrix/transport.py (hunk 5 above)
Comment (Dominik1999) [reply raiden/network/transport/matrix/transport.py]: ok, I always thought that checksum is a subset of hex :)

File: raiden/network/transport/matrix/utils.py (hunk 4 above)
Comment (ulope) [reply raiden/network/transport/matrix/utils.py]: Yeah also fine.

File: raiden/network/transport/matrix/transport.py (hunk 6 above)
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: > so it is intended to be an endless loop?

Yes, but only while the transport is running.

File: raiden/network/transport/matrix/transport.py (hunk 5 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: But there is a `ChecksumAddress` type, so you're right

File: raiden/network/transport/matrix/transport.py (hunk 1 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Makes perfect sense. I got confused by the logs, which were than only from the single call during start.

File: raiden/network/transport/matrix/transport.py (hunk 6 above)
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Okay so should I add to `transport.greenlets` or will it be cleared if the services main loop is shutdown?

Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Added them to transports greenlets.

File: raiden/network/transport/matrix/transport.py [hunk 10]
```diff
@@ -1159,27 +1178,27 @@ def _set_room_id_for_address(self, address: Address, room_id: Optional[_RoomID]
         If room_id is falsy, clean list of rooms. Else, push room_id to front of the list """
 
         assert not room_id or room_id in self._client.rooms, "Invalid room_id"
-        address_hex: AddressHex = to_checksum_address(address)
+        checksumaddress: ChecksumAddress = to_checksum_address(address)
```
[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: I'll go for `address_checksum`

File: raiden/network/transport/matrix/transport.py [hunk 11]
```diff
@@ -470,6 +471,24 @@ def on_success(greenlet):
...
+
+        def on_success(greenlet):
+            if greenlet in self.greenlets:
+                self.greenlets.remove(greenlet)
+
+        run_greenlet = gevent.spawn_later(0, func, *args, **kwargs)
+        run_greenlet.link_exception(self.on_error)
+        run_greenlet.link_value(on_success)
+
+        schedeule_greenlet = gevent.spawn_later(
```
[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Not sure I can follow, could you elaborate please?

 For me this approach has the benefit, that the executed function itself just takes the time it takes and the scheduling takes place in the event loop, depending on how busy it currently is.
 I had a while loop with sleep in it first, it produced timeout errors in many tests, as the `greenlets.get()` in `transport.stop()` was waiting for the sleep to finish. 
Also, why 15 seconds? I think this depends on how busy the main loop is, which makes sense to me.

File: raiden/network/transport/matrix/transport.py [hunk 12]
```diff
@@ -470,6 +471,24 @@ def on_success(greenlet):
         self.greenlets.append(greenlet)
         return greenlet
 
+    def _schedule(self, period: float, func: Callable, *args, **kwargs):
+        """Spawns two greenlets: one to call func with args immediately,
+        another one to call this function again after period has passed."""
+
+        def on_success(greenlet):
+            if greenlet in self.greenlets:
```
[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: That's very true..:)

[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: Done

File: raiden/network/transport/matrix/transport.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: 15s because on average the stop event will happen halfway through the 30s sleep (law of large numbers...)

Instead of a simple sleep you could use `gevent.wait(stop_event, 30)`. That way it will immediately return once the stop event is set.

[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: I misunderstood your first comment. 
The case you described is handled by transport.scheduled_greenlets.killall() in transport.stop() s.t. the parent does not have to wait for the scheduled children. I prefer the spawn_later approach, as we don‘t have to write an explicit while loop for every singular greenlet that’s supposed to run periodically. If we want to have greenlets that are not supposed to be killed on flight, there sure is a way to only kill those scheduled for later execution.

File: raiden/network/transport/matrix/transport.py [hunk 13]
```diff
@@ -470,6 +473,28 @@ def on_success(greenlet):
         self.greenlets.append(greenlet)
         return greenlet
 
+    def _schedule(self, period: float, func: Callable, *args, **kwargs):
+        """Spawns two greenlets: one to call func with args immediately,
+        another one to call this function again after period has passed."""
+        if self._stop_event.ready():
+            return
+
+        def on_success(greenlet):
```
Comment (palango) [inline raiden/network/transport/matrix/transport.py]: missing type annotation

File: raiden/network/transport/matrix/transport.py [hunk 14]
```diff
@@ -470,6 +473,28 @@ def on_success(greenlet):
...
+            if greenlet in self.scheduled_greenlets:
+                self.scheduled_greenlets.remove(greenlet)
+
+        run_greenlet = gevent.spawn_later(0, func, *args, **kwargs)
+        run_greenlet.link_exception(self.on_error)
+        run_greenlet.link_value(on_success)
+        self.scheduled_greenlets.append(run_greenlet)
+
+        schedeule_greenlet = gevent.spawn_later(
```
Comment (palango) [inline raiden/network/transport/matrix/transport.py]: s/schedeule_greenlet/schedule_greenlet

File: raiden/network/transport/matrix/transport.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: The question is will we ever have more than this one periodic thing?

Smell a bit [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) to me ;)

File: raiden/network/transport/matrix/transport.py [hunk 15]
```diff
@@ -1159,27 +1184,27 @@ def _set_room_id_for_address(self, address: Address, room_id: Optional[_RoomID]
         If room_id is falsy, clean list of rooms. Else, push room_id to front of the list """
 
         assert not room_id or room_id in self._client.rooms, "Invalid room_id"
-        address_hex: AddressHex = to_checksum_address(address)
+        address_checksum: ChecksumAddress = to_checksum_address(address)
```
Comment (palango) [inline raiden/network/transport/matrix/transport.py]: I think we should stick with `AddressHex` here, it's the type used everywhere in the client, even though it's nto the return type of `to_checksum_address`. But in the future we can make `AddressHex` an alias for `ChecksumAddress`.

File: raiden/tests/integration/network/transport/test_matrix_transport.py (hunk 9 above)
Comment (palango) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: Can you make this a constant, so that we don't have a central location to change these timeouts.

File: raiden/network/transport/matrix/transport.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (err508) [reply raiden/network/transport/matrix/transport.py]: For instance [#3262](https://github.com/raiden-network/raiden/issues/3262) is something that should be done/checked periodically i think

File: raiden/network/transport/matrix/transport.py [hunk 16]
```diff
@@ -395,6 +396,7 @@ def on_success(greenlet):
         self._client.sync_thread.link_exception(self.on_error)
         self._client.sync_thread.link_value(on_success)
         self.greenlets = [self._client.sync_thread]
+        self._schedule(30, self._address_mgr.get_address_mgr_info)
```
Comment (palango) [inline raiden/network/transport/matrix/transport.py]: Just a question for my understanding: why do we want to run this function twice, why only twice (compared to "every 30s), ond why after 30s?

Comment (err508) [reply raiden/network/transport/matrix/transport.py]: It's supposed to run every 30s, `_schedule` calls itself after 30s again

Comment (palango) [reply raiden/network/transport/matrix/transport.py]: Ah, my bad. Then I'm with Ulo in the bview that a while loop is easier to understand.

File: raiden/tests/integration/network/transport/test_matrix_transport.py [hunk 17]
```diff
@@ -142,7 +142,7 @@ def ping_pong_message_success(transport0, transport1):
     transport1._raiden_service.sign(ping_message)
     transport1.send_async(queueid0, ping_message)
 
-    with Timeout(20, exception=False):
+    with Timeout(40, exception=False):
```
Comment (palango) [inline raiden/tests/integration/network/transport/test_matrix_transport.py]: Can you make this a constant, so that we have a central location to change these timeouts.

Comment (err508) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: @palango I just discussed with @ulope that he'll take care of this in #4333

Comment (err508) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: So I would skip the additional CI run if there isn't anything else?

Comment (palango) [reply raiden/tests/integration/network/transport/test_matrix_transport.py]: Fine for me!
